components/Visualizer/Visualizer.tsx

```diff
@@ -24,38 +24,75 @@ const _Visualizer: FunctionComponent<VisualizerProps> = ({
   offScreenCanvasSupport
 }) => {
   const canvasRef = useRef<HTMLCanvasElement>(null);
+  const hiddenCanvasElement = useRef<HTMLCanvasElement>(
+    document.createElement("canvas")
+  ).current;
+
   const visualizerRef = useRef<HTMLDivElement>(null);
 
   const dimensions = useWindowResize(visualizerRef, {
     width: 1100,
     height: 400
   });
 
+  const isOffscreenSupported =
+    offScreenCanvasSupport === OFFSCREEN_2D_CANVAS_SUPPORT.SUPPORTED;
   useEffect(() => {
     (async function() {
-      const canvas: any =
-        offScreenCanvasSupport === OFFSCREEN_2D_CANVAS_SUPPORT.SUPPORTED
-          ? canvasRef.current.transferControlToOffscreen()
-          : canvasRef.current;
+      const canvas: any = false //isOffscreenSupported
+        ? canvasRef.current.transferControlToOffscreen()
+        : canvasRef.current;
+
+      const mainCanvasContext = (canvas as HTMLCanvasElement).getContext("2d");
 
       // This has been done because it wasn't getting correctly transferred
       // in firefox.
       const dimensions = JSON.parse(
         JSON.stringify(visualizerRef.current.getBoundingClientRect())
       );
-
-      await canvasProxy(
-        transfer(
-          {
-            canvas,
-            message: VISUALIZER_MESSAGES.INIT,
-            dimensions,
-            range,
-            mode
-          },
-          [canvas]
-        )
-      );
+      // TODO: mimicking no offscreencanvas support, remove later
+      if (false && isOffscreenSupported) {
+        await canvasProxy(
+          transfer(
+            {
+              canvas,
+              message: VISUALIZER_MESSAGES.INIT,
+              dimensions,
+              range,
+              mode
+            },
+            [canvas]
+          )
+        );
+      } else {
+        // else use hidden canvas
+        function paintToMainCanvas() {
+          mainCanvasContext.clearRect(
+            0,
+            0,
+            hiddenCanvasElement.width,
+            hiddenCanvasElement.height
+          );
+          mainCanvasContext.drawImage(
+            hiddenCanvasElement,
+            0,
+            0,
+            hiddenCanvasElement.width,
+            hiddenCanvasElement.height
+          );
+          // I think this is too agressive, for later, what's Clock?
```

Comment:
Clock keeps track of time and progress of the song. It's separate because it provides better API that way. For eg. when you need to pause and song on canvas, you just have to stop that clock. So it's king of virtual clock which is in our control. 
- If we stop it, the canvas pauses. 
- If we change the speed, the clock's speed should be increased.
- If we skip to a particular position, the clock skips to that note's time.

Eventually based on all these metrics, it provide the song progress(0-1) which is used by `Visualizer` utility to control the visualizations.

So if progress is moving faster, the canvas will be fast. If progress is stopped, the canvas remains at that position etc.

You can assume that the `Clock` actually is the time machine for the canvas which we can control.